sharedbuffers/mapped_struct.py

```diff
@@ -462,30 +492,43 @@ def is_wrapped_key(obj):
 @cython.inline
 def get_wrapped_key(obj):
     # the key for the unwrapped value
+    if six.PY2:
+        if isinstance(obj, tuple):
+            return obj[1]
+        elif isinstance(obj, int):
+            return obj
+        elif isinstance(obj, long):
+            return obj & ~WRAP_MASK
+
     if isinstance(obj, tuple):
         return obj[1]
     elif isinstance(obj, int):
-        return obj
-    elif isinstance(obj, long):
         return obj & ~WRAP_MASK
 
 
 class mapped_frozenset(frozenset):
 
     @classmethod
-    @cython.locals(cbuf = 'unsigned char[:]', i=int, ix=int, offs=cython.longlong)
+    @cython.locals(cbuf = 'unsigned char[:]', i=int, ix=int, offs=cython.longlong, cdtype=cython.uchar)
     def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
         if isinstance(obj, npndarray):
             all_int = all_intlong = all_float = 0
             obj_dtype = obj.dtype
             if obj_dtype.isbuiltin:
-                dtype = obj_dtype.char
-                cdtype = cython.cast('const char*', dtype)[0]
-                if cdtype in ('l', 'I', 'i', 'H', 'h', 'B', 'b'):
+                if six.PY3:
+                    dtype = obj_dtype.char.encode()
+                    if cython.compiled:
+                        cdtype = dtype[0]
+                    else:
+                        cdtype = bytes([dtype[0]])
+                else:
+                    dtype = obj_dtype.char
+                    cdtype = cython.cast('const char*', dtype)[0]
```

Comment:
You can't have this check (six.py3) in performance-critical code, as it compiles to a python attribute access on the global module. A relatively slow operation.

Store the result of this check into a cython.bint global variable, and use that variable instead if you need to.

Otherwise, try to write version-agnostic code.

In this case, if you declare dtype as unicode, `dtype[0] in ('l', 'I', 'i', ...)` generates pretty good code if you take a look. You would need 2 verisons of the code, py2 and py3, but the py3 version would actually be quite simpler (no char pointer stuff).